rclcpp/include/rclcpp/sched_base.hpp

```diff
@@ -0,0 +1,97 @@
+#pragma once
+
+#include <sched.h>
+#include <sys/syscall.h>
+#include <unistd.h>
+#include <stdint.h>
+#include <pthread.h>
+#include <atomic>
+
+
+namespace rclcpp {
+// forward declarations
+namespace executors {
+    class SingleThreadedExecutor;
+}; // rclcpp::executors
+namespace sched {
+
+struct pthread_struct {
+    void* __unused[90]; // Hacking here, may be platform dependent.
+    pid_t tid;
+};
+
+struct SchedAttr {
+    uint32_t size = sizeof(SchedAttr);              /* Size of this structure */
+    uint32_t sched_policy = SCHED_OTHER;      /* Policy (SCHED_*) */
+    uint64_t sched_flags;       /* Flags */
+    int32_t sched_nice;        /* Nice value (SCHED_OTHER, SCHED_BATCH) */
+    uint32_t sched_priority;    /* Static priority (SCHED_FIFO, SCHED_RR) */
+
+    /* For SCHED_DEADLINE */
+    uint64_t sched_runtime;
+    uint64_t sched_deadline;
+    uint64_t sched_period;
+
+    /* Utilization hints, unused for our purpose,
+       may enable in the future*/
+    // uint32_t sched_util_min;
+    // uint32_t sched_util_max;
+};
+
+/** Since pthread does not expose pid to us, this is a hack to get the (linux) pid.
+ *  This might be dangerous and non-portable.
+ */
+inline pid_t
+get_pid(pthread_t threadid) {
+    auto pthread_id = ((pthread_struct*) threadid);
+    /* this may occur if the thread is detached from the current thread, use this
+       function before calling detach() */
+    if (pthread_id == nullptr) {
+        printf("nullptr is passed to get_pid!\n");
+        return 0;
+    }
+    return pthread_id->tid;
+}
+
+/** Equality for SchedAttr, use it to prevent unnecessary syscall. */
+inline bool
+operator==(const SchedAttr& lhs, const SchedAttr& rhs) {
+    if (&lhs == &rhs) {
+        return true;
+    }
+    uint64_t* a = (uint64_t*) &lhs;
+    uint64_t* b = (uint64_t*) &rhs;
+    size_t offset = 0;
+    while (offset++ < (sizeof(SchedAttr) / 8)) {
```

Comment:
Why not use `return std::memcmp(&lhs, &rhs, sizeof(SchedAttr)) == 0;`? 

You're effectively breaking the SchedAttr struct objects into 8-byte chunks (uint64_t) and comparing them. Doing the comparison byte-by-byte would be logically equivalent. Using memcmp can do this for you, without using any syscalls. 

The reason I suggest this is not only to make the code simpler, but also because doing the comparisons in 8-byte chunks only works if you assume your struct's total size is padded to 8-byte increments. The current implementation of SchedAttr is 48-bytes, so this works. However, if the Linux kernel extends the struct in the future by adding something like another int32_t at the end, then your mirrored version of the struct here would become 52-bytes and the 8-byte comparison loop would start comparing 4 out-of-bound bytes.

Reply:
Addressed.